### PR TITLE
GODRIVER-2083: Test serialization of BSON with embedded null bytes in strings

### PR DESCRIPTION
**Pull request by gabbyasuncion**
[GODRIVER-2083](https://jira.mongodb.org/browse/GODRIVER-2083)

Syncs Go Driver BSON corpus spec tests with those indicated [here](https://github.com/mongodb/specifications/pull/1066/files). Also updates BSON corpus spec test to validate null bytes when encoding JSON to BSON and to raise the expected error (as done [here](https://github.com/mongodb/mongo-python-driver/pull/723/files#diff-5b2e28c0ea88c9ecfbb500a445ced611ab359dc70d4b64dc0788600b5803da3aR203-R206)).

### CODE REVIEW

File: bson/bson_corpus_spec_test.go [hunk 1]
```diff
@@ -350,6 +350,9 @@ func runTest(t *testing.T, file string) {
 					case "0x00", "0x05", "0x13":
 						var doc D
 						err := UnmarshalExtJSON([]byte(s), true, &doc)
+						if strings.Contains(p.Description, "Null") {
```
Comment (benjirewis) [inline bson/bson_corpus_spec_test.go:353]: Could you leave a comment here (for posterity) explaining why "Null" test-cases need to be `Marshal`ed to raise the correct error?

**Review by benjirewis — CHANGES_REQUESTED** [review]

**Review by benjirewis — APPROVED** [review]
LGTM! Thanks 👍

**Review by kevinAlbs — COMMENTED** [review]
Looks great!

I think one more case in the [prose tests](https://github.com/mongodb/specifications/commit/c64fc79899ed7f734db5a37d4b142ab6b82c097b#diff-6a4c00217c2bed9a55b113d415c5b1eac328c740c7289ea2499784be539f1536R364-R380) is required.

I suggest adding to the tests added in GODRIVER-1923:

- TestNullBytes in bson/marshal_test.go
- TestNullBytes in x/bsonx/bsoncore/bsoncore_test.go

Those have cases for embedded NULL bytes, but I think need the case `Field name within a sub-document`.

File: x/bsonx/bsoncore/bsoncore_test.go [hunk 2]
```diff
@@ -935,6 +935,12 @@ func TestNullBytes(t *testing.T) {
 			})
 		}
 	})
+	t.Run("sub document field name", func(t *testing.T) {
+		createDocFn := func() {
+			NewDocumentBuilder().StartDocument("a\x00").AppendDocument("foobar", []byte("foo")).FinishDocument()
```
Comment (benjirewis) [inline x/bsonx/bsoncore/bsoncore_test.go:940]:
```suggestion
			NewDocumentBuilder().StartDocument("foobar").AppendDocument("a\x00", []byte("foo")).FinishDocument()
```
I _think_ this is making a document where a top-level key contains null bytes. We want to be asserting that BSON creation panics in the document builder when null bytes are in a field that is embedded.

**Review by benjirewis — CHANGES_REQUESTED** [review]

File: x/bsonx/bsoncore/bsoncore_test.go (hunk 2 above)
Comment (gabbyasuncion) [reply x/bsonx/bsoncore/bsoncore_test.go:940]: Ah yeah that seems right thank you

**Review by benjirewis — APPROVED** [review]
LGTM! Sorry for not spotting the need for prose tests when I approved before.

**Review by kevinAlbs — APPROVED** [review]
LGTM
